rfd/0054-passwordless-macos.md

```diff
@@ -0,0 +1,419 @@
+
+* `tsh --pwdless_mode={auto,on,off}` - activate/deactivate passwordless
+  authentication
+
+    `auto` is the default behavior described above (passwordless is preferred
+    if we are certain it may be used)
+
+    `on` enables passwordless logins (`tsh login --pwdless`, described in the
+    [Passwordless FIDO2 RFD][passwordless fido2], is a shortcut to it)
+
+    `off` disables passwordless logins
+
+* `tsh --mfa_mode={auto,platform,cross-platform}` - choose whether to use
+  platform or cross-platform MFA
+
+    `auto` is the default behavior described above, which favors Touch ID
+
+    `platform` prefers platform authenticators, such as Touch ID, over OTP or
+    portable FIDO2 keys
+
+    `cross-platform` prefers FIDO2 or OTP (aka `tsh` behavior prior to this RFD)
+
+Both `--pwdless_mode` and `--mfa_mode` may be specified for fine-grained control
+over `tsh` authentication logic. Note that those flags apply not only to
+`tsh login`, but also for any commands that require re-authentication (such as
+`tsh mfa add`).
+
+Finally, if there are Touch ID credentials for multiple users and the login user
+is not known, `tsh login` may prompt the user to specify the `--user` flag.
+
+Example of a passwordless Touch ID login:
+
+```shell
+$ tsh login --proxy=example.com
+<system shows Touch ID prompt>
+> > Profile URL:        https://example.com
+>   Logged in as:       codingllama
+>   Cluster:            example.com
+>   Roles:              access, editor
+>   Logins:             codingllama
+>   Kubernetes:         enabled
+>   Valid until:        2021-10-04 23:32:29 -0700 PDT [valid for 12h0m0s]
+>   Extensions:         permit-agent-forwarding, permit-port-forwarding, permit-pty
+```
+
+### Detecting Touch ID support
+
+Detecting Touch ID support is important so `tsh` may enable/disable related
+features as appropriate.
+
+Apart from Go build tags, which are a rather coarse "detection" mechanism, we
+can take inspiration from [Chromium's implementation](
+https://github.com/chromium/chromium/blob/c4d3c31083a2e1481253ff2d24298a1dfe19c754/device/fido/mac/touch_id_context.mm#L126)
+and do the following checks:
+
+* Verify macOS version (>=10.12.2)
+* Verify if the `keychain-access-groups` entitlement is present
+* [LAContext canEvaluatePolicy:kLAPolicyDeviceOwnerAuthenticationWithBiometrics](
+  https://developer.apple.com/documentation/localauthentication/lacontext/1514149-canevaluatepolicy?language=objc) check
+* Secure Enclave check (attempt to create a key using
+  `kSecAttrIsPermanent = @NO`)
+
+<!--
+TODO(codingllama): In theory it's all fine, but let's run some tests on older
+ machines to be safe.
+-->
+
+### `tsh` support commands
+
+The following support commands are added to `tsh` as hidden subcommands. They
+are useful to diagnose and manage certain aspects of Touch ID support.
+
+The commands are only available on macOS builds.
+
+`tsh touchid diag` - prints diagnostics about Touch ID support (for example, if
```

Comment:
Great idea!

Reply:
I like my debug commands. :)

This one is borne of paranoia that a client somewhere will fail under mysterious reasons, so there's at least some data we can ask them for.